### PR TITLE
feat(replay): Add SentryMask and SentryUnmask iOS components

### PR DESCRIPTION
**Pull request by krystofwoldrich**
## :loudspeaker: Type of change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring

## :scroll: Description

This PR adds custom redact/mask components for iOS. 

The implementation includes noop fallback for Android and other unsupported platforms and fallback Web implementation. These will be added in a follow up PRs. 

```jsx
import * as Sentry from '@sentry/react-native'

const Example = () => (
  <View>
    <Sentry.Unmask>
      <Text>This is unmasked because it's direct child of Sentry.Unmask (can be masked if Sentry.Masked is used higher in the hierarchy)</Text>
      <Sentry.Mask>
        <Text>This is masked always because it's a child of a Sentry.Mask</Text>
        <Sentry.Unmask>
          {/* Sentry.Unmask does not override the Sentry.Mask from above in the hierarchy */}
          <Text>This is masked always because it's a child of Sentry.Mask</Text>
        </Sentry.Unmask>
      </Sentry.Mask>
    </Sentry.Unmask>
  </View>
);
```

| Device | Replay |
|--------|--------|
| Screenshot 2024-11-11 at 16 24 03 | Screenshot 2024-11-11 at 16 24 41 | 

```[tasklist]
### Todo
- [x] Requires https://github.com/getsentry/sentry-cocoa/releases/tag/8.40.0
- [ ] https://github.com/getsentry/sentry-cocoa/pull/4522
- [x] iOS Legacy Native Components
- [x] iOS New Arch Components
- [x] Android Legacy Components (in a follow up PR)
- [x] Android New Arch Components (in a follow up PR)
- [x] Web Component (in a follow up PR)
```

## :bulb: Motivation and Context

This is a part of RN Replay GA tasks 
- https://github.com/getsentry/sentry/issues/74441

## :green_heart: How did you test it?
ci, local,

## :pencil: Checklist

- [ ] I reviewed submitted code
- [ ] I added tests to verify changes
- [ ] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled
- [ ] All tests passing
- [ ] No breaking changes

### CODE REVIEW

File: samples/react-native/src/App.tsx [hunk 1]
```diff
@@ -112,7 +112,7 @@ Sentry.init({
   // otherwise they will not work.
   // release: 'myapp@1.2.3+1',
   // dist: `1`,
-  profilesSampleRate: 1.0,
+  // profilesSampleRate: 1.0,
```
Comment (krystofwoldrich) [inline samples/react-native/src/App.tsx:115]:
```suggestion
  profilesSampleRate: 1.0,
```

File: packages/core/ios/RNSentryReplay.m [hunk 2]
```diff
@@ -31,14 +32,25 @@ + (void)updateOptions:(NSMutableDictionary *)options
...
+    // To avoid typos, we test the class existens in the tests
+    return @[ @"RNSentryReplayUnmask" ];
+}
+
 + (NSArray *_Nonnull)getReplayRNRedactClasses:(NSDictionary *_Nullable)replayOptions
 {
     NSMutableArray *_Nonnull classesToRedact = [[NSMutableArray alloc] init];
+    // We can't import RNSentryReplayMask.h here because it's Objective-C++
+    // To avoid typos, we test the class existens in the tests
+    [classesToRedact addObject:@"RNSentryReplayMask"];
```
Comment (krystofwoldrich) [inline packages/core/ios/RNSentryReplay.m:53]: For Unmask to work this PR requires changes from https://github.com/getsentry/sentry-cocoa/pull/4522

Add add call of `PrivateSentrySDKOnly.setIgnoreWrapperClass(IgnoreWrapper.self)`

File: samples/react-native-macos/src/Screens/PlaygroundScreen.tsx [hunk 3]
```diff
@@ -32,6 +33,16 @@ const PlaygroundScreen = () => {
             <View style={styles.container}>
               <Text>Text:</Text>
               <Text>{'This is <Text>'}</Text>
+              <Text>Custom Mask:</Text>
+              <View>
+                {/* Replay is not supported on macOS, this test that the unavailable components do not crash the app */}
```
Comment (lucas-zimerman) [inline samples/react-native-macos/src/Screens/PlaygroundScreen.tsx:38]: suggestion: 
```suggestion
                {/* Replay is not supported on macOS. This sample demonstrates that unavailable components do not cause the app to crash. */}
```

File: packages/core/test/replay/CustomMask.test.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  it('returns a fallback when native view manager is missing', () => {
+    jest.mock('react-native', () => ({
+      UIManager: {},
+      View: jest.fn(),
+    }));
+
+    const { Mask, Unmask } = require('../../src/js/replay/CustomMask');
+
+    expect(Mask).toBeDefined();
+    expect(Unmask).toBeDefined();
```
Comment (lucas-zimerman) [inline packages/core/test/replay/CustomMask.test.ts:17]: Could we also validate that warn('Mask') / warn('Unask'); were called to make sure they are actually the fallback?
Same for the below tests with fallback.

File: packages/core/package.json [hunk 5]
```diff
@@ -118,7 +119,7 @@
   },
   "codegenConfig": {
     "name": "RNSentrySpec",
-    "type": "modules",
+    "type": "all",
```
Comment (lucas-zimerman) [inline packages/core/package.json:122]: Q: why this change?

File: packages/core/ios/RNSentryReplay.m [hunk 6]
```diff
@@ -36,6 +37,13 @@ + (void)updateOptions:(NSMutableDictionary *)options
                forKey:@"experimental"];
 }
 
++ (NSArray *_Nonnull)getReplayRNUnmaskClasses
+{
+    // We can't import RNSentryReplayUnmask.h here because it's Objective-C++
+    // To avoid typos, we test the class existens in the tests
```
Comment (lucas-zimerman) [inline packages/core/ios/RNSentryReplay.m:43]:
```suggestion
    // To avoid typos, we test the class existent in the tests
```

File: packages/core/package.json (hunk 5 above)
Comment (krystofwoldrich) [reply packages/core/package.json:122]: There are 3 types `all`, `modules` and `components`.

`modules` -> Native Module
`components` -> Native UI Components
`all` -> Combination of above.

File: CHANGELOG.md [hunk 7]
```diff
@@ -31,6 +31,25 @@
...
 
+- Add Replay Custom Masking for iOS ([#4238](https://github.com/getsentry/sentry-react-native/pull/4224))
+
+  ```jsx
+  import * as Sentry from '@sentry/react-native';
+
+  const Example = () => {
+    return (
+      <View>
+        <Sentry.Mask>
```
Comment (lucas-zimerman) [inline CHANGELOG.md:42]: Q: for now, <Sentry.Mask> / <Sentry.Unmask> are exclusive to native/ React Native?

File: packages/core/package.json (hunk 5 above)
Comment (lucas-zimerman) [reply packages/core/package.json:122]: Thank you for the explanation :)

File: packages/core/ios/RNSentryReplay.m [hunk 8]
```diff
@@ -56,10 +64,24 @@ + (NSArray *_Nonnull)getReplayRNRedactClasses:(NSDictionary *_Nullable)replayOpt
...
+    [PrivateSentrySDKOnly setRedactContainerClass:[self getMaskClass]];
+    [PrivateSentrySDKOnly setIgnoreContainerClass:[self getUnmaskClass]];
     RNSentryReplayBreadcrumbConverter *breadcrumbConverter =
         [[RNSentryReplayBreadcrumbConverter alloc] init];
     [PrivateSentrySDKOnly configureSessionReplayWith:breadcrumbConverter screenshotProvider:nil];
 }
 
++ (Class) getMaskClass
+{
+    return NSClassFromString(@"RNSentryReplayMask");
```
Comment (antonis) [inline packages/core/ios/RNSentryReplay.m:78]: `l`: This is a suggestion since the current approach works as expected and is covered by test cases.
Wdyt of using an objective-c wrapper to get the class names statically? Eg:

```
// RNSentryReplayWrapper.h

@interface RNSentryReplayWrapper : NSObject

+ (Class) getMaskClass;
+ (Class) getUnmaskClass

@end
```

```
// RNSentryReplayWrapper.mm

#import "RNSentryReplayWrapper.h"
#include "RNSentryReplayMask.h"
#include "RNSentryReplayUnmask.h"

@implementation MyObjectiveCWrapper

+ (Class)getMaskClass
{
    return RNSentryReplayMask.class;
}

+ (Class)getUnmaskClass
{
    return RNSentryReplayUnmask.class;
}

@end
```

File: packages/core/ios/RNSentryReplay.m (hunk 6 above)
Comment (krystofwoldrich) [reply packages/core/ios/RNSentryReplay.m:43]: Thanks.

File: CHANGELOG.md (hunk 7 above)
Comment (krystofwoldrich) [reply CHANGELOG.md:106]: This PR adds the component impl for iOS, next PR adds it for Android and next one for RN Web.

File: packages/core/ios/RNSentryReplay.m (hunk 8 above)
Comment (krystofwoldrich) [reply packages/core/ios/RNSentryReplay.m]: Maybe I've misunderstood but the wrapper it self would not help with the importing Cpp to Obj-C, but I've made different changes that enable us to use the classes directly.

https://github.com/getsentry/sentry-react-native/pull/4224/commits/050acade6e7d3e2d0843dbfbcb1e82e67a713c7b

**Comment by krystofwoldrich** [comment]
@antonis @lucas-zimerman This is ready for a review.

Since last review, the main change is in reverting shipping the TS source for RN Codegen and instead using NativeComponentsRegistry. https://github.com/getsentry/sentry-react-native/pull/4224/commits/eb457fc6eee542b8e09f571d2667b2711235dbe8

Since this will be the first time we ship new architecture compatible components, this will first be released as beta (likely 6.4.0-beta.1)

File: CHANGELOG.md [hunk 9]
```diff
@@ -8,6 +8,27 @@
 
 ## Unreleased
 
+### Features
+
+- Add Replay Custom Masking for iOS ([#4238](https://github.com/getsentry/sentry-react-native/pull/4224))
```
Comment (lucas-zimerman) [inline CHANGELOG.md:13]: NIT: PR number is different
```suggestion
- Add Replay Custom Masking for iOS ([#4224](https://github.com/getsentry/sentry-react-native/pull/4224))
```

File: packages/core/ios/RNSentryReplayBreadcrumbConverterHelper.m [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+
++ (void)configureSessionReplayWithConverter
+{
+    RNSentryReplayBreadcrumbConverter *breadcrumbConverter =
+        [[RNSentryReplayBreadcrumbConverter alloc] init];
+    [PrivateSentrySDKOnly configureSessionReplayWith:breadcrumbConverter screenshotProvider:nil];
+}
+
+@end
+#endif
```
Comment (lucas-zimerman) [inline packages/core/ios/RNSentryReplayBreadcrumbConverterHelper.m:16]: i: based on your other files, you are adding an extra line after `@end`
```suggestion
@end

#endif
```

**Comment by lucas-zimerman** [comment]
Lint seems to not be working

**Review by lucas-zimerman — COMMENTED** [review]
Overall, the PR looks good! but before merging, build/lint tests should pass, unless the issue is unrelated to the PR

File: packages/core/src/js/replay/CustomMask.tsx [hunk 11]
```diff
@@ -0,0 +1,63 @@
+import { logger } from '@sentry/utils';
+import * as React from 'react';
+import type { HostComponent, ViewProps } from 'react-native';
+import { UIManager, View, requireNativeComponent } from 'react-native';
```
Comment (antonis) [inline packages/core/src/js/replay/CustomMask.tsx:4]: NIT: Removing the unused import should fix the lint issue.

```suggestion
import { UIManager, View } from 'react-native';
```

**Review by antonis — APPROVED** [review]
Apart from [the lint issue](https://github.com/getsentry/sentry-react-native/pull/4224#discussion_r1856531071) the PR LGTM and works as expected 🎉

**Review by lucas-zimerman — APPROVED** [review]
LGTM!
